### PR TITLE
Spherical Matrix

### PR DESCRIPTION
**Pull request by zhanglw0521**
Spherical matrix implementation

### CODE REVIEW

File: src/properties.jl [hunk 1]
```diff
@@ -116,3 +118,37 @@ filter(φ::SphericalVector, b::Array) = ( length(b) <= 1 ? true :
...
 rot3Dcoeffs(::SphericalVector, T::DataType=Float64) = Rot3DCoeffs(T)
+
+
+struct SphericalMatrix{L1, L2, LEN1, LEN2, T} <: AbstractProperty
+   val::SMatrix{LEN1, LEN2, T}
+   _valL1::Val{L1}
+   _valL2::Val{L2}
+end
+
+getL(φ::SphericalMatrix{L1,L2}) where {L1,L2} = [L1;L2]
```
Comment (cortner) [inline src/properties.jl:129]: please return a tuple here: `... where {L1, L2} = L1, L2`

* `[L1; L2]` is actually a matrix; to construct a vector use `[L1, L2]`. 
* Such a vector is allocating, whereas a tuple is on the stack. Further, the values `L1, L2` can be inferred if they are returned as a tuple.

File: test/test_symmbasis.jl [hunk 2]
```diff
@@ -73,7 +73,58 @@ for L = 0 : 3
...
 end
 
-##
+#---
+@info("SymmetricBasis construction and evaluation: Spherical Matrix")
+
+for L1 = 0:3
+   for L2 = 0:3
+
+      if L1 == 0
```
Comment (cortner) [inline test/test_symmbasis.jl:82]: this is extremely ugly, why not define a function that gives you the symbol associated with each `L`? e.g., something along the lines of
```julia
const __L2syms = [:s, :p, :d, :f, :g, :h, :i, :k]
const __syms2L = Dict( [sym => L-1 for (L, sym) in enumerate(__L2syms)]... )
get_orbsym(L::Integer)  = __Lsyms[L+1]
get_L(sym::Union{AbstractString, Char, Symbol}) = __syms2L[ Symbol(sym) ]
```
should go somewhere inside the ACE module, not entirely sure yet where.

Comment (cortner) [reply test/test_symmbasis.jl:82]: and don't worry with the test numbers please

File: test/test_symmbasis.jl [hunk 3]
```diff
@@ -73,7 +73,58 @@ for L = 0 : 3
...
+         end
+      end
+
+      φ = ACE.SphericalMatrix(L1, L2; T = ComplexF64)
+      pibasis = PIBasis(B1p, ord, maxdeg; property = φ, isreal = false)
+      basis = SymmetricBasis(pibasis, φ)
+      BB = evaluate(basis, cfg)
+
+      for ntest = 1:5
+         Q, D1, D2 = ACE.Wigner.rand_QD(L1, L2)
```
Comment (cortner) [inline test/test_symmbasis.jl:114]: * Can you please fix `rand_QD` to do the following: return an *orthogonal* matrix `Q` with the sign already integrated, and modified D-matrices `D1, D2` with the `(-1)^L1` already integrated into `D1`.

Comment (cortner) [reply test/test_symmbasis.jl:114]: So the usage should end up 
```julia
Q, D1, D2 = ACE.Wigner.rand_QD(L1, L2)
Xs1 = shuffle(Ref(Q) .* Xs)
D1txBB1xD2 = Ref(D1') .* BB1 .* Ref(D2)
```
I.e. no more `rand_ref`

Comment (cortner) [reply test/test_symmbasis.jl:114]: please fix this also in the other test

**Review by cortner — CHANGES_REQUESTED** [review]
looks really good. But too much duplication now, let's get back to discussing how to unify everything. 

What about a `SphericalTensor`? Would that be any more difficult to write than what you already have?

**Review by cortner — APPROVED** [review]

**Review by cortner — APPROVED** [review]

**Comment by cortner** [comment]
looks great, thank you for your work on this.

**Comment by cortner** [comment]
@zhanglw0521 I've squased those commits into a single one and to make sure you don't get clashes I've deleted your branch. You can restore it if you will, but best to pull and checkout `main` now and then create a new `lz/equiv` branch.

**Comment by zhanglw0521** [comment]
> @zhanglw0521 I've squased those commits into a single one and to make sure you don't get clashes I've deleted your branch. You can restore it if you will, but best to pull and checkout `main` now and then create a new `lz/equiv` branch.

Seems that I'd better create a new one - will just follow your suggestion. Thanks a lot!

**Pull request merged**